blog/2022-06-21-using-temporal-as-a-node-task-queue.md

```diff
@@ -0,0 +1,372 @@
+
+```javascript
+const Queue = require('bull');
+const axios = require('axios');
+
+const emailQueue = new Queue('email');
+const mailgunAPI = 'https://api.mailgun.net/v3';
+
+queue.process(async function sendEmail(job) {
+  const { to, from, subject, html } = job.data;
+  await axios({
+    url: `${mailgunAPI}/${domain}/messages`,
+    method: 'post',
+    params: { to, from, subject, html },
+    auth: {
+      username: 'api',
+      password: apiKey
+    }
+  })
+});
+```
+
+`create-test-task.js`
+
+```javascript
+const Queue = require('bull');
+
+const emailQueue = new Queue('email');
+
+emailQueue.add({
+  to: 'john.smith@gmail.com',
+  from: 'test@temporal.io',
+  subject: 'Welcome to Temporal!',
+  html: 'This is a test email'
+});
+```
+
+With Temporal, the setup is similar.
+A task or job corresponds to a [Workflow](https://docs.temporal.io/concepts/what-is-a-workflow-execution).
+But, first, you need to define an [Activity](https://docs.temporal.io/concepts/what-is-an-activity) to send the email as follows.
+
+`src/activities.ts`
+
+```ts
+import axios from 'axios';
+import { SendEmailParams } from './interfaces';
+
+interface MailgunSettings {
+  apiKey: string;
+  domain: string;
+}
+
+const mailgunAPI = 'https://api.mailgun.net/v3';
+
+export const createActivities = ({ apiKey, domain }: MailgunSettings) => ({
+  async sendEmail(params: SendEmailParams): Promise<void> {
+    const { to, from, subject, html } = params;
+    await axios({
+      url: `${mailgunAPI}/${domain}/messages`,
+      method: 'post',
+      params: { to, from, subject, html },
+      auth: {
+        username: 'api',
+        password: apiKey,
+      },
+    });
+  },
+});
+```
+
+Next, you need to define a Workflow that calls this activity.
+
+`src/workflows.ts`
+
+```ts
+import { proxyActivities } from '@temporalio/workflow';
+import type { createActivities } from './activities';
+import { SendEmailParams } from './interfaces';
+
+const { sendEmail } = proxyActivities<ReturnType<typeof createActivities>>({
+  startToCloseTimeout: '5m',
+});
+
+export async function sendEmailWorkflow(params: SendEmailParams): Promise<void> {
+  await sendEmail(params);
+}
+```
+
+Finally, you need to create [Worker](https://docs.temporal.io/concepts/what-is-a-worker-entity) to process tasks as follows.
+
+`src/worker.ts`
+
+```ts
+import { Worker } from '@temporalio/worker';
+import dotenv from 'dotenv';
+import { createActivities } from './activities';
+
+dotenv.config();
+
+const {
+  MAILGUN_API: apiKey,
+  MAILGUN_DOMAIN: domain
+} = process.env;
+
+async function run(): Promise<void> {
+  if (apiKey === undefined) throw new Error('No API key');
+  if (domain === undefined) throw new Error('No domain');
+
+  const activities = createActivities({
+    apiKey,
+    domain,
+  });
+
+  const worker = await Worker.create({
+    taskQueue: 'email',
+    activities,
+    workflowsPath: require.resolve('./workflows'),
+  });
+  await worker.run();
+}
+
+run().catch((err) => {
```

Comment:
FTR, if you compile TS to esm you can use top level await and get rid of this wrapper.
We have an esm sample in the samples repo if you're interested in changing - up to you.